distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering/Natural_Order.enso

```diff
@@ -17,62 +20,66 @@ import Standard.Base.Data.Ordering.Vector_Lexicographic_Order
-        True ->
-            ## TODO [RW] Currently there is no `Integer.parse` method, so we
+    iter1 = BreakIterator.getCharacterInstance
+    iter1.setText text1
+
+    iter2 = BreakIterator.getCharacterInstance
+    iter2.setText text2
+
+    is_digit=character -> character>=48 && character<=57
+
+    get_number text prev next iter =
+        find_number text next iter =
+            new_next = iter.next
+            if (new_next == -1) then -next else
+                substring = Text_Utils.substring text next new_next
+                character = Text_Utils.get_chars substring . at 0
+                if (is_digit character).not then next else
+                    @Tail_Call find_number text new_next iter
+
+        end_index = find_number text next iter
+        substring = Text_Utils.substring text prev n.abs
+        
+        ## TODO [RW] Currently there is no `Integer.parse` method, so we
                parse a decimal and convert it to an integer. Once
                https://www.pivotaltracker.com/story/show/181176522 is
-               implemented, this should be changed to use `Integer.parse`.
-            value = Decimal.parse part . floor
-            Natural_Ordering_Numeric_Part value part
-        False ->
-            Natural_Ordering_Alphabetical_Part part
-
-## PRIVATE
-type Natural_Ordering_Part
-    ## PRIVATE
-    type Natural_Ordering_Alphabetical_Part (text : Text)
-
-    ## PRIVATE
-    type Natural_Ordering_Numeric_Part (value : Integer) (original_text : Text)
-
-    ## PRIVATE
-    compare_to : Natural_Ordering_Part -> Ordering
-    compare_to that = case Pair this that of
-        Pair (Natural_Ordering_Alphabetical_Part text1) (Natural_Ordering_Alphabetical_Part text2) ->
-            text1 . compare_to text2
-        Pair (Natural_Ordering_Numeric_Part value1 original1) (Natural_Ordering_Numeric_Part value2 original2) ->
-            value_ordering = value1.compare_to value2
-            case value_ordering of
-                Ordering.Equal -> original1 . compare_to original2
-                _ -> value_ordering
-        Pair (Natural_Ordering_Numeric_Part _ _) (Natural_Ordering_Alphabetical_Part _) ->
-            Ordering.Less
-        Pair (Natural_Ordering_Alphabetical_Part _) (Natural_Ordering_Numeric_Part _ _) ->
-            Ordering.Greater
-
-    to_text : Text
-    to_text = case this of
-        Natural_Ordering_Alphabetical_Part text -> text.to_text
-        Natural_Ordering_Numeric_Part v o -> [v, o].to_text
+               implemented, this should be changed to use `Integer.parse`.        
+        decimal = Decimal.parse substring . floor
+
+        next_index = if end_index < 0 then -1 else iter.current
+        [substring, decimal, end_index, next_index]
+
+
+    ## Ordering: Nothing < Number < Text
+    order prev1 next1 prev2 next2 =
+        case (next1 == -1) of
+            True ->
+                if (next2 == -1) then Ordering.Equal else Ordering.Less
+            False ->
```

Comment:
For such paired checks like this one, I think we can use nested pattern matching (IIRC it should work well currently) with Pairs, I think it can be a bit clearer.
```suggestion
    order prev1 next1 prev2 next2 =
        case Pair (next1 == -1) (next2 == -1) of
            Pair True True -> Ordering.Equal
            Pair True False -> Ordering.Less
            ...
```